automation/Extension/firefox/lib/http-post-parser.js

```diff
@@ -0,0 +1,250 @@
+      obj[key] = value;
+    });
+    return JSON.stringify(obj);
+  }catch (e) {
+    console.log("Exception: Cannot parse POST data:", e, encodingType, "formData:", formData);
+    return formData;  // return the original body if we can't decode
+  }
+}
+
+
+HttpPostParser.prototype.parsePostRequest = function(encodingType){
+  try {
+    this.parseStream();
+  } catch (e) {
+    console.log( "Exception: Failed to parse POST", e );
+    return {};
+  }
+
+  var postBody = this.postBody;
+
+  if (!postBody)  // some scripts strangely sends empty post bodies (confirmed with the developer tools)
+    return {};
+
+  var isMultiPart = false;  // encType: multipart/form-data
+  var postHeaders = this.postHeaders;  // request headers from upload stream
+  // See, http://stackoverflow.com/questions/16548517/what-is-request-headers-from-upload-stream
+
+  // add encodingType from postHeaders if it's missing
+  if (!encodingType && postHeaders && ("Content-Type" in postHeaders))
+    encodingType = postHeaders["Content-Type"];
+
+  if (encodingType.indexOf("multipart/form-data") != -1)
+    isMultiPart = true;
+
+  var jsonPostData = "";
+  var escapedJsonPostData = "";
+  if (isMultiPart) {
+    jsonPostData = this.parseMultiPartData(postBody, encodingType);
+    escapedJsonPostData = loggingDB.escapeString(jsonPostData);
+  } else {
+    jsonPostData = this.parseEncodedFormData(postBody, encodingType);
+    escapedJsonPostData = loggingDB.escapeString(jsonPostData);
+  }
+  return {post_headers: postHeaders, post_body: escapedJsonPostData};
+};
+
+
+
+HttpPostParser.prototype.parseMultiPartData = function(formData, encodingType) {
+  /*
+   * Parse POST bodies with encType "multipart/form-data encoded"
+   * https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#multipartform-data
+   *
+   * formData is in the following form:
+   *
+    -----------------------------12972102761018453617355621459
+    Content-Disposition: form-data; name="email"
+
+    test@example.com
+    -----------------------------12972102761018453617355621459
+    Content-Disposition: form-data; name="username"
+
+    name surname+
+    -----------------------------12972102761018453617355621459--
+    */
+  var boundary = "";
+  var firstLine = formData.split("\r\n", 1)[0];
+  if (firstLine.startsWith("-----------------------------"))
+    boundary = firstLine;
+  else
+    return formData;  // return unparsed data, if we fail to find the boundary string
+
+  var formVars = {};
+  for (var part of formData.split(boundary)){
+    partData = this.parseSinglePart(part);
+    if ("key" in partData && "value" in partData)
+      formVars[partData["key"]] = partData["value"];
+  }
+  return JSON.stringify(formVars);
+}
+
+
+HttpPostParser.prototype.parseSinglePart = function(part) {
+  /*
+   * Parse a single part of a multipart request body
+   * e.g., one part is as follows:
+     Content-Disposition: form-data; name="username"
+
+     name surname+
+  */
```

Comment:
According to: https://stackoverflow.com/questions/16548517/what-is-request-headers-from-upload-stream%60 there may be additional headers within parts, such as `Content-Type: `. From what I can tell your code doesn't handle this?